src/celestia/qt/qtglwidget.cpp

```diff
@@ -445,14 +445,25 @@ void CelestiaGlWidget::keyPressEvent( QKeyEvent* e )
     if (e->modifiers() & ControlModifier)
         modifiers |= CelestiaCore::ControlKey;
 
-    switch (e->key())
+    int ch = e->key();
```

Comment:
do we really need this part?

Reply:
We can remove this and change the if to:
```
-    int ch = e->key();
+    if ((e->key() >= 'B' && e->key() <= 'Y'))
+    {
+        if (e->modifiers().testFlag(Qt::ControlModifier))
+            appCore->charEntered(e->text().toUtf8().data(), modifiers);   // [Ctrl+ch]
+        else if (e->modifiers().testFlag(Qt::ShiftModifier))
+            appCore->charEntered(static_cast<char>(e->key()));        // toupper(ch)
+        else
+            appCore->charEntered(static_cast<char>(tolower(e->key())));

```
But the compiler should create the same binary code.

Reply:
I meant not the single line `int ch = e->key();` but the whole chunk
```+    if ((e->key() >= 'B' && e->key() <= 'Y'))
+    {
+        if (e->modifiers().testFlag(Qt::ControlModifier))
+            appCore->charEntered(e->text().toUtf8().data(), modifiers);   // [Ctrl+ch]
+        else if (e->modifiers().testFlag(Qt::ShiftModifier))
+            appCore->charEntered(static_cast<char>(e->key()));        // toupper(ch)
+        else
+            appCore->charEntered(static_cast<char>(tolower(e->key())));```

What doesn't work with it and how your code fixes it?

Other issues: why don't you check A & Z(*)? How this code behaves if both CapsLock & Shift are pressed?

If your additions are required then you have added them into the wrong location: they should be in `default` section of `switch` so you don't need to check for 'B' < key 'Y'.

* - I know the answer.

And question for the whole team: should we distinguish lower- & uppercase letters in this case?

Reply:
The VK_B (0x42) is a uppercase 'B' and there is no VK_A or VK_Z, so I check for 'B' to 'Y'

To check previous with if before switch, is much faster than in default. for the default, the CPU must check first all cases before go to default mark.
The default handle first the special keys with function handleSpecialKey(), there is another switch case for detecting cursor and other keys.
switch blocks have high complexity when you check with a code analyser.

The lower case is checked the lines before
    if (e->modifiers() & ShiftModifier)
        modifiers |= CelestiaCore::ShiftKey;

Ah, ok, I see in the handleSpecialKey() the key 'A' and 'Z' is also checked in case and call later the function:   appCore->charEntered(e->text().toUtf8().data(), modifiers);

I think the lower/uppercase letters calculated from e->text().toUtf8().data() is much overhead to convert each latin letter to UTF-8
I saw, the key 'A' and 'Z' must be handled in handleSpecialKey(), when I change from 'A' to 'Z' the speed can not be changed - 'A' increase speed, 'Z' decreased speed.

Reply:
You haven't answered the main question: What doesn't work without your addition and how it fixes that?